docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+
+Therefore, we will not allow Pods from the same Job to differ arbitrarily
+(anyway, users can use multiple Job objects for that case).  We will try to
+allow as little as possible to differ between pods of the same Job, while
+still allowing users to express common parallel patterns easily For users
+who need to run jobs which differ in other ways, they can create multiple
+Jobs, and manage them as a group using labels.
+
+
+From the above work lists, we see a need for Pods which differ in their command
+lines, and in their environment variables.  These work lists do not require the
+pods to differ in other ways.
+
+Experience in a [similar systems](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43438.pdf) has shown this model to be applicable
+to a very broad range of problems, despite this restriction.
+
+Therefore we will allow pods in the same Job to differ **only** in the following aspects:
+
+- command line
+- environment variables
+
+That is a statements about what the Job resource schema can express, and what
+the Job controller will do when creating pods of the same job.
+
+An admission controller could still modify pods, and some fields of pods could still
+be updated after creation.  Nothing will prevent this.
+Pods of the same Job will not have different volumes mounted on them, though
+users can mount all needed volumes and then select different files of interest.
+
+#### Interaction with vertical autoscaling
+
+Since pods of the same job will not be created with different resources,
+a vertical autoscaler will need to:
+
+- if it has index-specific initial resource suggestions, suggest those at admission
+time; it will need to understand indexes.
+- mutate resource requests on already created pods based on usage trend or previous container failures
+- modify the job template, affecting all indexes.
+
+### Composition of existing images
+
+The docker image that is used in a job may not be maintained by the person
+running the job.  Over time, the Dockerfile may change the ENTRYPOINT or CMD.
+If we require people to specify the complete command line to use Indexed Job,
+then they will not automatically pick up changes in the default
+command or args.
+
+This needs more thought.
+
+### Running Ad-Hoc Jobs using kubectl
+
+A user should be able to easily start an Indexed Job using `kubectl`.
+For example to run [work list 1](#work-list-1), a user should be able
+to type something simple like:
+
+```
+kubectl run process-files --image=myfileprocessor \
+   --per-completion-env=F="12342.dat 97283.dat 38732.dat" \
+   --restart=OnFailure  \
+   -- \
+   /usr/local/bin/process_file '$F'
+```
+
+In the above example:
+
+- `--restart=OnFailure` implies creating a job instead of replicationController.
+- Each pods command line is `/usr/local/bin/process_file $F`.
+- `--per-completion-env=` implies the jobs `.spec.completions` is set to the length of the argument array (3 in the example).
```

Comment:
You mean that if any per-completion-env is specified then completions is set by its length?